src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+    // formulasOfB := assertedFormulas - pFormulasOfA
+    Set<Term> formulasOfB =
+        assertedFormulas.stream()
+            .flatMap(c -> c.stream())
+            .filter(n -> !pFormulasOfA.contains(n))
+            .collect(ImmutableSet.toImmutableSet());
+
+    if (formulasOfB.isEmpty()) { // Catch trivial case
+      return mgr.getBooleanFormulaManager().makeBoolean(false);
+    }
+
+    // fit the Input to work with getCVC5Interpolation
+
+    ImmutableList<Collection<Term>> formAAsList = ImmutableList.of(pFormulasOfA);
+    ImmutableList<Collection<Term>> formBAsList = ImmutableList.of(formulasOfB);
+
+    Term itp = getCVC5Interpolation(formAAsList, formBAsList);
+
+    return creator.encapsulateBoolean(itp);
+  }
+
+  @Override
+  public List<BooleanFormula> getSeqInterpolants(
+      List<? extends Collection<Term>> partitionedFormulas)
+      throws SolverException, InterruptedException {
+    Preconditions.checkArgument(
+        !partitionedFormulas.isEmpty(), "at least one partition should be available.");
+
+    List<BooleanFormula> itps = Stream.of(new BooleanFormula[] {}).collect(Collectors.toList());
+    for (int i = 1; i < partitionedFormulas.size(); i++) {
+      itps.add(
+          getInterpolant(
+              ImmutableList.copyOf(Iterables.concat(partitionedFormulas.subList(0, i)))));
+    }
+    return itps;
+  }
+
+  // Experimental!
+  @Override
+  public List<BooleanFormula> getTreeInterpolants(
+      List<? extends Collection<Term>> pPartitionedFormulas, int[] pStartOfSubTree)
+      throws SolverException, InterruptedException {
+    Preconditions.checkState(!closed);
+    assert InterpolatingProverEnvironment.checkTreeStructure(
+        pPartitionedFormulas.size(), pStartOfSubTree);
+
+    // Generate every Interpolation Pair
+    ImmutableList<ImmutableList<ImmutableList<Collection<Term>>>> interpolationPairs =
+        getTreeInterpolationPairs(pPartitionedFormulas, pStartOfSubTree);
+
+    final List<Term> itps = Stream.of(new Term[] {}).collect(Collectors.toList());
+    try { // Interpolate every Interpolation Pair
+      for (int i = 0; i < interpolationPairs.size(); i++) {
+        itps.add(
+            getCVC5Interpolation(
+                interpolationPairs.get(i).get(0), interpolationPairs.get(i).get(1)));
+      }
+    } catch (UnsupportedOperationException e) {
+      if (e.getMessage() != null) {
+        throw new SolverException(e.getMessage(), e);
+      } else {
+        throw e;
+      }
+    }
+
+    final List<BooleanFormula> result =
+        Stream.of(new BooleanFormula[] {}).collect(Collectors.toList());
+    for (Term itp : itps) {
+      result.add(creator.encapsulateBoolean(itp));
+    }
+    assert result.size() == pStartOfSubTree.length - 1;
+    return result;
+  }
+
+  /**
+   * Interpolates a Tuple of Interpolants according to Craig-Interpolation using CVC5-Interpolation.
+   *
+   * <p>CVC5's getInterpolant: There is a Model, such that the Interpolant I, with A -> I = True
```

Comment:
I am still missing the third constraint on interpolants: which symbols appear in the interpolant?

Reply:
CVC5s documentation does not mention which symbols appear in the interpolant. Is there a elegant way to compare appearing symbols between formulas/terms?

Reply:
The FormulaManager can collect variables and UFs from a formula with calling `fmgr.extractVariablesAndUFs(formula)`.

Reply:
I added a small check after each interpolation, but only checking for the maps keys. Is it sufficient?